File: rxspecialpermissions/src/main/java/com/ivianuu/rxspecialpermissions/permission/PermissionBuilder.java
/*
 * Copyright 2017 Manuel Wrage
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.ivianuu.rxspecialpermissions.permission;

import android.content.Context;
import android.graphics.drawable.Drawable;
import android.support.annotation.DrawableRes;
import android.support.annotation.NonNull;
import android.support.annotation.RestrictTo;
import android.support.annotation.StringRes;
import android.support.v4.content.ContextCompat;

import com.ivianuu.rxspecialpermissions.Permission;

/**
 * Real permission builder
 */
public final class PermissionBuilder {

    private final Context context;

    private String title;
    private String description;
    private Drawable icon;
    private RealPermission.GrantedProvider grantedProvider;
    private RealPermission.IntentProvider intentProvider;

    @RestrictTo(RestrictTo.Scope.LIBRARY)
    public PermissionBuilder(@NonNull Context context) {
        this.context = context.getApplicationContext();
    }

    /**
     * Sets the title of this permission
     */
    @NonNull
    public PermissionBuilder titleRes(@StringRes int titleRes) {
        return title(context.getString(titleRes));
    }

    /**
     * Sets the title of this permission
     */
    @NonNull
    public PermissionBuilder title(@NonNull String title) {
        this.title = title;
        return this;
    }

    /**
     * Sets the description of this permission
     */
    @NonNull
    public PermissionBuilder descriptionRes(@StringRes int descriptionRes) {
        return description(context.getString(descriptionRes));
    }

    /**
     * Sets the description of this permission
     */
    @NonNull
    public PermissionBuilder description(@NonNull String description) {
        this.description = description;
        return this;
    }

    /**
     * Sets the icon of this permission
     */
    @NonNull
    public PermissionBuilder iconRes(@DrawableRes int iconRes) {
        return icon(ContextCompat.getDrawable(context, iconRes));
    }

    /**
     * Sets the icon of this permission
     */
    @NonNull
    public PermissionBuilder icon(@NonNull Drawable icon) {
        this.icon = icon;
        return this;
    }

    /**
     * Sets the granted predicate
     */
    @NonNull
    public PermissionBuilder grantedProvider(@NonNull RealPermission.GrantedProvider grantedProvider) {
        this.grantedProvider = grantedProvider;
        return this;
    }

    /**
     * Sets the intent provider
     */
    @NonNull
    public PermissionBuilder intentProvider(@NonNull RealPermission.IntentProvider intentProvider) {
        this.intentProvider = intentProvider;
        return this;
    }

    /**
     * Returns the permission
     */
    @NonNull
    public Permission build() {
        if (title == null) {
            throw new IllegalStateException("missing title");
        }
        if (grantedProvider == null) {
            throw new IllegalStateException("missing granted provider");
        }
        if (intentProvider == null) {
            throw new IllegalStateException("missing intent provider");
        }

        return new RealPermission(title, description, icon, grantedProvider, intentProvider);
    }
}

File: rxspecialpermissions/src/main/java/com/ivianuu/rxspecialpermissions/permission/PermissionBuilder.kt
/*
 * Copyright 2017 Manuel Wrage
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *  
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.ivianuu.rxspecialpermissions.permission

import android.content.Context
import com.ivianuu.rxspecialpermissions.delegate.*

/**
 * Factory for [RealPermissionBuilder]'s
 */
object PermissionBuilder {

    @JvmStatic
    fun accessibility(
        context: Context,
        clazz: Class<*>
    ): RealPermissionBuilder = withDelegate(context, AccessibilityDelegate(context, clazz))

    @JvmOverloads
    @JvmStatic
    fun deviceAdmin(
        context: Context,
        clazz: Class<*>,
        explanation: String? = null
    ): RealPermissionBuilder = withDelegate(context, DeviceAdminDelegate(context, clazz, explanation))

    @JvmStatic
    fun ignoreBatteryOptimizations(
        context: Context
    ): RealPermissionBuilder = withDelegate(context, IgnoreBatteryOptimizationsDelegate(context))

    @JvmStatic
    fun installUnknownApps(
        context: Context
    ): RealPermissionBuilder = withDelegate(context, InstallUnknownAppsDelegate(context))

    @JvmStatic
    fun notificationListener(
        context: Context,
        clazz: Class<*>
    ): RealPermissionBuilder = withDelegate(context, NotificationListenerDelegate(context, clazz))

    @JvmStatic
    fun notificationPolicyAccess(
        context: Context
    ): RealPermissionBuilder = withDelegate(context, NotificationPolicyAccessDelegate(context))

    @JvmStatic
    fun packageUsageStats(
        context: Context
    ): RealPermissionBuilder = withDelegate(context, PackageUsageStatsDelegate(context))

    @JvmStatic
    fun systemOverlay(
        context: Context
    ): RealPermissionBuilder = withDelegate(context, SystemOverlayDelegate(context))

    @JvmStatic
    fun unrestrictedDataAccess(
        context: Context
    ): RealPermissionBuilder = withDelegate(context, UnrestrictedDataAccessDelegate(context))

    @JvmStatic
    fun vrListener(
        context: Context,
        clazz: Class<*>
    ): RealPermissionBuilder = withDelegate(context, VrListenerDelegate(context, clazz))

    @JvmStatic
    fun writeSettings(
        context: Context
    ): RealPermissionBuilder = withDelegate(context, WriteSettingsDelegate(context))

    @JvmStatic
    fun withDelegate(
        context: Context,
        delegate: RealPermission.Delegate
    ): RealPermissionBuilder =
        RealPermissionBuilder(context, delegate)
}